File: Data Structures & Algorithms/house-robber/submission-4.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0] * (n+2)
        for i in range(2, n+2):
            dp[i] = max(nums[i]+dp[i-2], d[i-1])
        return dp[-1]

File: Data Structures & Algorithms/house-robber-ii/submission-4.cpp
class Solution {
private:
    int f(vector<int>& nums, int i, int j) {
        int minusTwo = 0;
        int minusOne = 0;

        for(int x = i; x <= j; x++) {
            int cur = max(minusTwo + nums[x], minusOne);
            minusTwo = minusOne;
            minusOne = cur;
        }

        return minusOne;
    }
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        return max(f(nums, 1, n-1), f(nums, 0, n-2));
    }
};
